collector/textfile.go

```diff
@@ -0,0 +1,125 @@
+// +build !notextfile
+
+package collector
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	dto "github.com/prometheus/client_model/go"
+
+	"code.google.com/p/goprotobuf/proto"
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/text"
+)
+
+var (
+	textFileDirectory = flag.String("textfile.directory", "", "Directory to read text files with metrics from.")
+)
+
+type textFileCollector struct {
+}
+
+func init() {
+	Factories["textfile"] = NewTextFileCollector
+}
+
+// Takes a config struct and registers a
+// SetMetricFamilyInjectionHook.
+func NewTextFileCollector(config Config) (Collector, error) {
+	if *textFileDirectory == "" {
+		// This collector is enabled by default, so do not fail if
+		// the flag is not passed.
+		glog.Infof("No directory specified, see --textfile.directory")
+	} else {
+		prometheus.SetMetricFamilyInjectionHook(parseTextFiles)
+	}
+
+	return &textFileCollector{}, nil
+}
+
+// textFile collector works via SetMetricFamilyInjectionHook in parseTextFiles.
+func (c *textFileCollector) Update(ch chan<- prometheus.Metric) (err error) {
+	return nil
+}
+
+func parseTextFiles() []*dto.MetricFamily {
+	var parser text.Parser
+	error := 0.0
+	metricFamilies := make([]*dto.MetricFamily, 0)
+	mtimes := map[string]time.Time{}
+
+  // Iterator over files and accumulate their metrics.
+	files, _ := ioutil.ReadDir(*textFileDirectory)
+	for _, f := range files {
+		if !strings.HasSuffix(f.Name(), ".prom") {
```

Comment:
I'd considered that, to be correct you'd need to escape the passed in directory and you'd also have to parse out the results to get the file name (I don't want the full path in the mtime metric, as that'd make it harder to write rules as the directory could change.)